File: Data Structures & Algorithms/course-schedule-ii/submission-1.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # DFS with cycle detection
        ans = []
        preq = defaultdict(list)
        visited = set() # fully taken
        curr = set()    # current dfs path
        for crs, pre in prerequisites:
            preq[crs].append(pre)

        for i in range(numCourses):
            if not self.dfs(i, preq, curr, visited, ans):
                return []
        
        return ans
    
    def dfs(self, crs, preq, curr, visited, ans):
        if crs in visited:
            return True
        if crs in curr:
            return False
        
        curr.add(crs)
        for p in preq[crs]:
            if not self.dfs(p, preq, curr, visited, ans):
                return False
        ans.append(crs)
        visited.add(crs)
        curr.remove(crs)
        return True
        

File: Data Structures & Algorithms/decode-ways/submission-1.cpp
class Solution {
public:
    int numDecodings(string s) {
        int ans = 0;
        std::vector<int> dp(s.size(), -1);
        dfs(s, 0, ans, dp);
        return dp[0] > 0 ? dp[0] : 0;
    }

    int dfs(string& s, int start, int& ans, std::vector<int>& dp) {
        if (start >= s.size()) {
            return 1;
        }

        if (s[start] == '0') {
            return 0;
        }

        if (dp[start] != -1) {
            return dp[start];
        }

        int ways = dfs(s, start + 1, ans, dp);
        if (start < s.size() - 1 && 
                (s[start] == '1' || (s[start] == '2' && s[start+1] >= '0' && s[start+1] <= '6'))) {
            ways += dfs(s, start + 2, ans, dp);
        }
        dp[start] = ways;
        return ways;
    }
};
